### PR TITLE
Require in Section: warning is now about fragility not deprecation.

### PR DESCRIPTION
**Pull request by SkySkimmer**
Depends on https://github.com/coq/coq/pull/11972

### CODE REVIEW

**Comment by Zimmi48** [comment]
This warning is documented in the vernacular commands chapter. This should be updated (removed).

**Comment by ppedrot** [comment]
FTR, I am opposed to this. Even if this now works in a sound way in the kernel, this kind of feature is messing with other invariants in the libobject layer. (I actually believe that even the kernel code is far from optimal, since it requires separate storage of monomorphic constraints coming from requires-in-sections.)

**Comment by ejgallego** [comment]
I don't have a particular strong opinion on this one, however I don't like that our deprecation criteria should be based on our lack of technical ability to make things work.

If `Require` inside sections is useful for users then we should fix the implementation to that's done in a proper way, not remove the feature. In principle, I don't see a strong technical reason why Require inside a section should not work.

**Comment by Zimmi48** [comment]
I also dislike the use of deprecation to signal broken things but not followed by any removal action. If there is something in particular, that is broken, the warning could say so (it wouldn't be a deprecation warning). If there is something fundamental that cannot be fixed, then we should proceed with the removal. But if there is nothing fundamental nor any clear bug that we know of, then we should simply remove the deprecation.

**Comment by ppedrot** [comment]
> if there is nothing fundamental nor any clear bug that we know of

There is a third reason: enforcing a sane document model. In the end, barring undecidability, there is rarely any fundamental reason not to be able to implement something, but it's not necessarily the thing to do.

**Comment by Zimmi48** [comment]
What is unsane with `Require` in sections? We are not talking about `Require` in proofs, where side-effects are known to be bad. We are talking about the fact that at some point in the document a library is required, when it was not the case before. There can be good reasons for not wanting to put some `Require` at the top of a document. Example:

```coq
Module Omega.

Ltac omega := idtac.

End Omega.

Print Omega.omega.

Require ZArith.

Print Omega.omega.
```

There is no reason why the fact that this `Require` command is in the middle of a section should change anything.

**Comment by ejgallego** [comment]
I am a bit with @Zimmi48 in that I don't see a particular conceptual problem with `Require` in the middle of a document; if we had a bit saner semantics in terms of state management would be almost a non-issue.

**Comment by ppedrot** [comment]
> What is unsane with Require in sections?

Merely adding to the combinatorial complexity of libobjects in a pain in itself. If you have an object that has some require-scoped effect, then now you need to be aware of that and think twice when replaying the effect at section closure time. Given that require-bound effects tend to be low-level imperative changes, this is asking for trouble without a solid reason. Also @maximedenes has probably something to say on that topic.

If people want to write this nonetheless, fine with me, but then we shouldn't be surprised to get a flood of bugs related the use of requires in sections.

**Comment by ejgallego** [comment]
I agree that we should at the very least not encourage it given the status of the implementation.

But deprecation must go with care given the feedback we've got about this feature being important for some users.

**Comment by ppedrot** [comment]
I suspect that the was confusion with another ugly feature (probably nested proofs) because otherwise people would have realized this was broken by now and would have complained. Before @SkySkimmer's PR, the feature was virtually not usable, so it's probably not used.

**Comment by Zimmi48** [comment]
>If you have an object that has some require-scoped effect, then now you need to be aware of that and think twice when replaying the effect at section closure time.

But we want to remove these require-scoped effects, right? When we do finish removing them, would there still be an issue with `Require` in sections?

>If people want to write this nonetheless, fine with me, but then we shouldn't be surprised to get a flood of bugs related the use of requires in sections.

I wouldn't mind that we change the deprecation warning into a "this feature is known to be broken" warning (that is not the same thing at all!). Especially, if it seems that at some point in the future, there are good chances that all the bugs will have been solved.

>But deprecation must go with care given the feedback we've got about this feature being important for some users.

I am not aware of this feedback.

**Comment by SkySkimmer** [comment]
>But we want to remove these require-scoped effects, right?

I'm not sure about that.

>I wouldn't mind that we change the deprecation warning into a "this feature is known to be broken"  warning

It's not "known to be broken" if we don't have an example of it breaking.

**Comment by Zimmi48** [comment]
> It's not "known to be broken" if we don't have an example of it breaking.

Indeed. Then, we could use a less strong language like "fragile". OTOH, maybe this flood of bug reports would be a good thing because it would help to figure out what parts of the implementation are fragile and should be improved.

**Comment by ejgallego** [comment]
> "this feature is known to be broken"

I still think my rejected PR adding "KNOWN_ISSUES" was a good idea.

> I am not aware of this feedback.

Was some time ago in Coq Club, but basically MIT-style long sessions do find out that they need to Require stuff in the middle of a document when doing proofs, redoing all the proofs to that point in some cases can take days.

**Comment by Zimmi48** [comment]
>Was some time ago in Coq Club, but basically MIT-style long sessions do find out that they need to Require stuff in the middle of a document when doing proofs, redoing all the proofs to that point in some cases can take days.

Right! This is in fact the same UI issue that justify the need for nested proofs.

**Comment by Zimmi48** [comment]
> I still think my rejected PR adding "KNOWN_ISSUES" was a good idea.

I still keep this idea in mind but with the idea of putting it inside the reference manual instead of as a separate file.

**Comment by SkySkimmer** [comment]
I weakened this PR: the warning changed a bit to avoid implyin that we plan to remove the functionality but we still warn.

**Review by Zimmi48 — APPROVED** [review]
I am supportive of this change.

File: vernac/vernacentries.ml [hunk 1]
```diff
@@ -1066,10 +1066,9 @@ let vernac_end_segment ({v=id} as lid) =
 (* Libraries *)
 
 let warn_require_in_section =
-  let name = "require-in-section" in
-  let category = "deprecated" in
-  CWarnings.create ~name ~category
-    (fun () -> strbrk "Use of “Require” inside a section is deprecated.")
+  CWarnings.create ~name:"require-in-section" ~category:"fragile"
+    (fun () -> strbrk "Use of “Require” inside a section is fragile." ++ spc() ++
+               strbrk "Committing files which use this functionality is not recommended.")
```
Comment (ejgallego) [inline vernac/vernacentries.ml:1071]: I'd suggest; "using this functionality in production-ready Coq code is not recommended"

Comment (ejgallego) [reply vernac/vernacentries.ml:1071]: Still the warning, while better now, it is not very informative; how could we improve this without adding a very long story?

**Review by ejgallego — APPROVED** [review]
I support this tweak too; it seems like it should target 8.12, WDYT @Zimmi48 ?

File: vernac/vernacentries.ml (hunk 1 above)
Comment (Zimmi48) [reply vernac/vernacentries.ml:1071]: I would say "in finished proof scripts" to highlight the fact that this can be useful in the middle of a work.

**Comment by Zimmi48** [comment]
Yes, it's fine since it's a user message.

**Comment by ejgallego** [comment]
@ppedrot WDYT ? Are you OK with the current wording, would you like it to be stronger?

Deprecation is not feasible in the short term IMVHO, IMO it is better to tweak the message for now; note that this PR doesn't make an statement about whether the feature will remain supported, neither does commit to not deprecating it eventually.

**Comment by ppedrot** [comment]
Wording is fine, but this won't prevent me from believing we're actively looking for trouble. I'd be less concerned if this feature was removed altogether.

**Comment by ejgallego** [comment]
> Wording is fine, but this won't prevent me from believing we're actively looking for trouble. I'd be less concerned if this feature was removed altogether.

Main blocker is [this comment](https://github.com/coq/coq/pull/10473#pullrequestreview-410231609) , my understanding from feedback from users is that at the moment it would quite disruptive.

Indeed it would be nice if we could simplify stuff in the lines of #10473 , but it doesn't seem feasible to me in the short term.

@SkySkimmer , do you want to tweak the message a bit then so we ship this in the beta?

**Comment by Zimmi48** [comment]
> do you want to tweak the message a bit then so we ship this in the beta?

Note that it is fine if this PR waits a day or two because this qualifies for backporting.

**Comment by Zimmi48** [comment]
@SkySkimmer Did you see Emilio's comment about the wording of the message? Do you want to tweak it?

**Comment by ejgallego** [comment]
@Zimmi48 , seems @SkySkimmer may be busy; would you like to tweak the message yourself, otherwise I guess we can merge.

**Pull request merged**